python_modules/dagster/dagster/_core/execution/asset_backfill.py

```diff
@@ -927,6 +929,63 @@ def _check_validity_and_deserialize_asset_backfill_data(
     return asset_backfill_data
 
 
+def backfill_is_complete(
+    backfill_id: str, backfill_data: AssetBackfillData, instance: DagsterInstance
+):
+    """A backfill is complete when:
+    1. all asset partitions in the target subset have a materialization state (successful, failed, downstream of a failed partition).
+    2. there are no in progress runs for the backfill.
+    3. there are no failed runs that will result in an automatic retry.
+
+    Condition 1 ensures that for each asset partition we have attempted to materialize it or have determined we
+    cannot materialize it because of a failed dependency. Condition 2 ensures that no retries of failed runs are
+    in progress. Condition 3 guards against a race condition where a failed run could be automatically retried
+    but it was not added into the queue in time to be caught by condition 2.
+
+    Since the AssetBackfillData object stores materialization states per asset partition, we want to ensure the
+    daemon continues to update the backfill data until all runs have finished in order to display the
+    final partition statuses in the UI.
+    """
+    from dagster._daemon.auto_run_reexecution.auto_run_reexecution import (
+        filter_runs_to_should_retry,
+    )
+
+    # conditions are in order of least expensive to most expensive to check
+    if (
+        # Condition 1 - all asset partitions in the target subset have a materialization state
+        backfill_data.all_targeted_partitions_have_materialization_status()
+        # Condtition 2 - there are no in progress runs for the backfill
+        and len(
+            instance.get_run_ids(
+                filters=RunsFilter(
+                    statuses=NOT_FINISHED_STATUSES,
+                    tags={BACKFILL_ID_TAG: backfill_id},
+                ),
+                limit=1,
+            )
+        )
+        == 0
+        # Condition 3 - there are no failed runs that will be retried
+        and len(
+            list(
+                filter_runs_to_should_retry(
```

Comment:
I'm a little worried about the performance this function call. `filter_runs_to_should_retry` involved fetching the run group for each run that is passed to the function. Is there a different/better way to find out if a run will be retried?

Reply:
This is the airlift PR I mentioned running into similar challenges with making decisions based on whether or not a run will actually retry: https://github.com/dagster-io/dagster/pull/25761

Reply:
We should delete that `dagster/run_will_retry` tag after the retry kicks off, right? So that a `get_runs` call filtering on that tag will only fetch runs that are about to retry, and not runs that have already retried

Reply:
Given that we can add run tags after a run has been created, I don't see a reason why we can't delete a run tag after creation as well

Reply:
working on the retry changes here https://github.com/dagster-io/dagster/pull/25932 still in draft as i work things out, but we can move discussion around how the tag should work to this PR

Reply:
I think we can replace the run_will_retry tag with a child_run_id or a retried_run_id tag. Making it efficient for cases where we only have a graphql response of the original run I think is ideal

Reply:
I don't think there are any current places in the codebase that delete run tags - i wonder if there's a way to calculate this from looking at the full set of failed runs without needing to delete any

Reply:
Isn't will_retry plus the presence of a child run tag enough? You can just follow the run chain to its conclusion

Reply:
A run could have multiple child runs though which is tricky to stick in a singular key value pair

Reply:
what's the case in which that happens exactly? If it's just the fact that a run may retry multiple times I think it's fine to list only the depth 1 retry and rely on successive calls to get to the end of the chain.

Reply:
I don't think there's a case where a run would have multiple automatic retries, but it could have multiple manual retries or one automatic one manual. 

With the scheme as is in https://github.com/dagster-io/dagster/pull/25932 the `dagster/child_run` tag would only be added for automatic retries. but that could also cause issues later on if we wanted to follow the chain of child runs down for any retry (manual or automatic)

Reply:
Manual retries already don't count towards your retry limit though right? So I think it's kinda irrelevant?

Reply:
my hypothesis is that my manual retry is completing before the automatic retry system processes the original run failure. it sees `run 1` as part of the run group and so sets the retry number from there. but the same thing would happen if a manual retry finished between any of the retried runs

Reply:
to close the loop on this based on our convo in slack - we're gonna add a `did_retry` tag once the retry has been kicked off